[SciPy][Specfun]: Divergence in `test_exponential_integrals.py::TestExp1::test_branch_cut`

Responsible subroutine: `scipy/special/specfun.f`

```fortran
        SUBROUTINE E1Z(Z,CE1)
C
C       ====================================================
C       Purpose: Compute complex exponential integral E1(z)
C       Input :  z   --- Argument of E1(z)
C       Output:  CE1 --- E1(z)
C       ====================================================
C
```

Test logs:

```console
$ python dev.py test -t scipy.special.tests.test_exponential_integrals -v
....
============================================= FAILURES =============================================
_____________________________________ TestExp1.test_branch_cut _____________________________________
scipy/special/tests/test_exponential_integrals.py:12: in test_branch_cut
    assert sc.exp1(complex(-1, 0)).imag == (
E   AssertionError: assert -3.141592653589793 == --3.141592653589793
E    +  where -3.141592653589793 = (-1.8951178163559368-3.141592653589793j).imag
E    +    where (-1.8951178163559368-3.141592653589793j) = <ufunc 'exp1'>((-1+0j))
E    +      where <ufunc 'exp1'> = sc.exp1
E    +      and   (-1+0j) = complex(-1, 0)
E    +  and   -3.141592653589793 = (-1.8951178163559368-3.141592653589793j).imag
E    +    where (-1.8951178163559368-3.141592653589793j) = <ufunc 'exp1'>((-1-0j))
E    +      where <ufunc 'exp1'> = sc.exp1
E    +      and   (-1-0j) = complex(-1, -0.0)
        self       = <scipy.special.tests.test_exponential_integrals.TestExp1 object at 0x7fe95c22ed10>
===================================== short test summary info ======================================
FAILED scipy/special/tests/test_exponential_integrals.py::TestExp1::test_branch_cut - AssertionError: assert -3.141592653589793 == --3.141592653589793
```

Reproducible code:

```fortran
subroutine e1z(z)
   implicit complex*16 (c,z)
   implicit double precision (a,d-h,o-y)
   pi=3.141592653589793d0
   el=0.5772156649015328d0
   x=dble(z)
   a0=abs(z)
   xt=-2*dabs(dimag(z))
   if (a0.eq.0.0d0) then
      ce1=(1.0d+300,0.0d0)
   else if (a0.le.5.0.or.x.lt.xt.and.a0.lt.40.0) then
      ce1=(1.0d0,0.0d0)
      cr=(1.0d0,0.0d0)
      do 10 k=1,500
         cr=-cr*k*z/(k+1.0d0)**2
         ce1=ce1+cr
         if (abs(cr).le.abs(ce1)*1.0d-15) go to 15
10    continue
15    continue
      if (x.le.0.0.and.dimag(z).eq.0.0) then
         ce1=-el-log(-z)+z*ce1-sign(pi,dimag(z))*(0.0d0,1.0d0)
      else
         ce1=-el-log(z)+z*ce1
      endif
   else
      zc=0d0
      zd=1/z
      zdc=1*zd
      zc=zc + zdc
      do 20 k=1,500
         zd=1/(zd*k + 1)
         zdc=(1*zd - 1)*zdc
         zc=zc + zdc

         zd=1/(zd*k + z)
         zdc=(z*zd - 1)*zdc
         zc=zc + zdc

         if (abs(zdc).le.abs(zc)*1.0d-15.and.k.gt.20) go to 25
20    continue
25    ce1=exp(-z)*zc
      if (x.le.0.0.and.dimag(z).eq.0.0) ce1=ce1-pi*(0.0d0,1.0d0)
   endif

   print *, ce1
   return
end

program main
    complex*16 z
    z = (-1.0d0, 0.0d0)
    call e1z(z)
    z = (-1.0d0, -0.0d0)
    call e1z(z)
end program

```

> GFortran

```console
$ gfortran ./a.f90 && ./a.out
             (-1.8951178163559368,-3.1415926535897931)
              (-1.8951178163559368,3.1415926535897931)
```

> LFortran

```console
$ lfortran a.f90 --implicit-typing --no-warnings
(-1.895118,-3.141593)
(-1.895118,-3.141593)
```

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -481,6 +481,8 @@ RUN(NAME intrinsics_60 LABELS gfortran llvm COPY_TO_BIN intrinsics_60_data.txt) 
 RUN(NAME intrinsics_61 LABELS gfortran llvm) # char
 RUN(NAME intrinsics_62 LABELS gfortran llvm) # shape
 RUN(NAME intrinsics_63 LABELS gfortran llvm)
+RUN(NAME intrinsics_64 LABELS gfortran llvm wasm) # sign
+
 
 RUN(NAME parameter_01 LABELS gfortran)
 RUN(NAME parameter_02 LABELS gfortran)

--- a/integration_tests/intrinsics_64.f90
+++ b/integration_tests/intrinsics_64.f90
@@ -1,0 +1,9 @@
+program intrinsics_64
+    real :: x, y
+    x = 1.0
+    y = -0.0
+
+    x = sign(x, y)
+    print *, x
+    if (abs(x - (-1.0)) > 1e-9) error stop
+end program

--- a/src/libasr/ASR.asdl
+++ b/src/libasr/ASR.asdl
@@ -264,6 +264,7 @@ expr
     | RealUnaryMinus(expr arg, ttype type, expr? value)
     | RealCompare(expr left, cmpop op, expr right, ttype type, expr? value)
     | RealBinOp(expr left, binop op, expr right, ttype type, expr? value)
+    | RealCopySign(expr target, expr source, ttype type, expr? value)
     | ComplexConstant(float re, float im, ttype type)
     | ComplexUnaryMinus(expr arg, ttype type, expr? value)
     | ComplexCompare(expr left, cmpop op, expr right, ttype type, expr? value)

--- a/src/libasr/codegen/asr_to_llvm.cpp
+++ b/src/libasr/codegen/asr_to_llvm.cpp
@@ -5326,6 +5326,37 @@ public:
         tmp = lfortran_str_slice(str, left, right, step, left_present, right_present);
     }
 
+    void visit_RealCopySign(const ASR::RealCopySign_t& x) {
+        if (x.m_value) {
+            this->visit_expr_wrapper(x.m_value, true);
+            return;
+        }
+        this->visit_expr(*x.m_target);
+        llvm::Value* target = tmp;
+
+        this->visit_expr(*x.m_source);
+        llvm::Value* source = tmp;
+
+        llvm::Type *type;
+        int a_kind;
+        a_kind = down_cast<ASR::Real_t>(ASRUtils::type_get_past_pointer(x.m_type))->m_kind;
+        type = llvm_utils->getFPType(a_kind);
+        llvm::Value *ftarget = builder->CreateSIToFP(target,
+                type);
+        llvm::Value *fsource = builder->CreateSIToFP(source,
+                type);
+        std::string func_name = a_kind == 4 ? "llvm.copysign.f32" : "llvm.copysign.f64";
+        llvm::Function *fn_copysign = module->getFunction(func_name);
+        if (!fn_copysign) {
+            llvm::FunctionType *function_type = llvm::FunctionType::get(
+                    type, { type, type}, false);
+            fn_copysign = llvm::Function::Create(function_type,
+                    llvm::Function::ExternalLinkage, func_name,
+                    module.get());
+        }
+        tmp = builder->CreateCall(fn_copysign, {ftarget, fsource});
+    }
+
     template <typename T>
     void handle_SU_IntegerBinOp(const T &x) {
         if (x.m_value) {

--- a/src/libasr/codegen/asr_to_wasm.cpp
+++ b/src/libasr/codegen/asr_to_wasm.cpp
@@ -1484,6 +1484,24 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
         }
     }
 
+    void visit_RealCopySign(const ASR::RealCopySign_t& x) {
+        if (x.m_value) {
+            visit_expr(*x.m_value);
+            return;
+        }
+        this->visit_expr(*x.m_target);
+        this->visit_expr(*x.m_source);
+
+        int kind = ASRUtils::extract_kind_from_ttype_t(x.m_type);
+        if (kind == 4) {
+            m_wa.emit_f32_copysign();
+        } else if (kind == 8) {
+            m_wa.emit_f64_copysign();
+        } else {
+            throw CodeGenError("visit_RealCopySign: Only kind 4 and 8 reals supported");
+        }
+    }
+
     void visit_RealBinOp(const ASR::RealBinOp_t &x) {
         if (x.m_value) {
             visit_expr(*x.m_value);

--- a/src/libasr/pass/intrinsic_function_registry.h
+++ b/src/libasr/pass/intrinsic_function_registry.h
@@ -1231,23 +1231,18 @@ namespace Sign {
         fill_func_arg("x", arg_types[0]);
         fill_func_arg("y", arg_types[0]);
         auto result = declare(fn_name, return_type, ReturnVar);
-        /*
-         * r = abs(x)
-         * if (y < 0) then
-         *     r = -r
-         * end if
-        */
         if (is_real(*arg_types[0])) {
-            ASR::expr_t *zero = f(0, arg_types[0]);
-            body.push_back(al, b.If(fGtE(args[0], zero), {
-                b.Assignment(result, args[0])
-            }, /* else */ {
-                b.Assignment(result, f32_neg(args[0], arg_types[0]))
-            }));
-            body.push_back(al, b.If(fLt(args[1], zero), {
-                b.Assignment(result, f32_neg(result, arg_types[0]))
-            }, {}));
+            Vec<ASR::expr_t*> args; args.reserve(al, 2);
+            visit_expr_list(al, new_args, args);
+            ASR::expr_t* real_copy_sign = ASRUtils::EXPR(ASR::make_RealCopySign_t(al, loc, args[0], args[1], arg_types[0], nullptr));
+            return real_copy_sign;
         } else {
+            /*
+            * r = abs(x)
+            * if (y < 0) then
+            *     r = -r
+            * end if
+            */
             ASR::expr_t *zero = i(0, arg_types[0]);
             body.push_back(al, b.If(iGtE(args[0], zero), {
                 b.Assignment(result, args[0])
@@ -1257,12 +1252,12 @@ namespace Sign {
             body.push_back(al, b.If(iLt(args[1], zero), {
                 b.Assignment(result, i32_neg(result, arg_types[0]))
             }, {}));
-        }
 
-        ASR::symbol_t *f_sym = make_Function_t(fn_name, fn_symtab, dep, args,
-            body, result, Source, Implementation, nullptr);
-        scope->add_symbol(fn_name, f_sym);
-        return b.Call(f_sym, new_args, return_type, nullptr);
+            ASR::symbol_t *f_sym = make_Function_t(fn_name, fn_symtab, dep, args,
+                body, result, Source, Implementation, nullptr);
+            scope->add_symbol(fn_name, f_sym);
+            return b.Call(f_sym, new_args, return_type, nullptr);
+        }
     }
 
 } // namespace Sign

